### PR TITLE
[PORT] Full Heretic Overhaul

### PR DESCRIPTION
**Pull request by Ivniinvi**
## About The Pull Request
Ports and partial ports from the following PRs:

- tgstation/tgstation#54252
- tgstation/tgstation#64658
- tgstation/tgstation#34633
- tgstation/tgstation#62918
- tgstation/tgstation#60496
- tgstation/tgstation#60947
- tgstation/tgstation#63789
- tgstation/tgstation#59977
- tgstation/tgstation#56233
- tgstation/tgstation#52287

Gameplay changes:

- The heretic no longer starts with a Codex Cicatrix or a Living Heart.

- Heretics now draw transmutation runes by using any writing tool while having Mansus Grasp active.
- The Mansus Grasp can be used to remove heretic runes.
- While draining influences, people who examine you may get a message hinting that you're interacting with an influence.
- Drained influences can also be dispersed with anomaly neutralizers!
- The Codex Cicatrix is now a researchable item that lets you gain additional knowledge from influences.
- The Codex can still draw and remove runes, and does it faster.
- The Living Heart is now the heretic's heart. Literally. It's the heart in their chest. Their heart takes on the appearance of a living heart, and they can pulse it on demand to track their targets. This makes an audible noise.
- If your heart is lost (you're disemboweled or whatever), you need to do a ritual to regain it!
- Casting any heretic spell (besides Grasp) requires a Focus.

- A Heretic Focus is a neck item they can transmute.
- Heretic robes also function as a focus when toggled up.
- Ascending also disables the need for a focus, because of course.
- Heretics now gain 1 knowledge point passively every 20 minutes.

- Sacrificing has been revamped entirely.

- A heretic now gains four sacrifice targets on spawn.
- One random crewmember
- One member of their department
- One member of security
- One member of command (a "high value" sacrifice)
- You can sacrifice people who are in in soft-crit, hard-crit, or dead.
- Sacrificing someone will cuff them (if they are not), HEAL them, revive them, and set them unconscious. After a short time. they will be sent to a shadow realm. This shadow realm is themed to your heretic type.
- The shadow realm is a 2 minute long survival challenge where the sacrificee is subject to a constant barrage of shadowy hands.
- If they survive, they are teleported back to a random turf with no memory of how they got there. They'll also slur a TON to get the point across.
- If they die, their corpse is teleported back to a random turf on the station.
- No more multi-hearting! Your targets are your own.

- BUT adds a knowledge that allows heretics to reroll their sacrifice targets with a ritual.
- Each path now has a "Rituals of Knowledge". These are randomly generated rituals that may be difficult to complete but awards knowledge points in return.

- Ascending now has some requirements.

- To learn the ascension ritual, you need to complete all of the objective you are assigned.
- The ascension ritual now each have a varied requirement, instead of "needing 3 bodies" only.
- Other minor gameplay changes:

- Lots of balance tweaking.
- Buffed some summons.
- Buffed the Lord of the Night very slightly.
- Nerfed the Madness Mask.
- Put a limit on the amount of blade transmutations possible at once. 3 for flesh, 2 for other paths.
- Logs of BUG fixing.
- General grammar and flavor tweaks all around.
- Admin / code changes:
- Revamped the way heretics appear within the traitor panel.
- You can now easily see who they're targeting for sacrifice and what they have researched
- Also adds some helpful buttons to heretics, like giving them points!
- Refactored much, much of heretic code
- LIKE ALL OF HERETIC CODE WAS IN 4 FILES.
- Split up all the knowledge, spells, and items that belong to the heretic into their own files and folders.
- Not only that, but everything internally was still named "Eldritch Cultist" and similar.
- Almost every mention of "Eldritch Cultist" has been properly replaced by "Heretic".
- Much better reference handling all around.
- General code improvements over heretic stuff.
- Unit tests, because of course.

## Why It's Good For The Game

Heretics are not in a good place at the moment, this much is clear. They've been completely disabled on MRP for this reason.

The reasoning is simple: A lot of murder.
There's nothing inherently wrong with an antagonist heavy with murder, but the Heretic really missed the mark.
Gib, gib, gib, then ascend so you can keep killing.

In the background, the Heretic was FULL of flavorful spells, rituals, and "lore" stolen from Cultist Simulator that was unfortunate enough to be shackled to the heretic's gameplay loop.

So, this revamp aims to amend that:

Dial back the heretic's focus on mass murder and put more focus on the heretic's interesting flavor.
Spooky maintenance rituals, knowledge seeking maniac.

Sacrifice no longer outright kills / requires murder, meaning a heretic can progress without racking up a bodycount.
Influence is gained passively over time, so they can spend influence on more interesting side paths.
Side paths are required to progress to ascension, so they're encouraged to explore new things.
Ultimately, while there still may be a little way to go, this PR seeks to take a good leap in starting it.

## Testing Procedure

I did a bunch of testing over the past two days and I'm reasonably confident that the bulk of it is working as intended, I do think that this should definitely be testmerged before a fullmerge.

## Changelog
:cl: Ivniinvi, EdgeLordExe, Melbert, Imaginos, SpaceManiac, ChungusGamer666, AnturK, ZephyrTFA, coiax, Krysonism
add: New heretic path - Path of Void- it specializes in being extremely stealthy.
tweak : Removed curse of blindness replaced with mask of madness. hey it even rhymes.
add: Large scale heretic revamp!
add: The Codex Cicatrix is no longer a roundstart heretic item. Research is handled through an action button. Rune drawing is done by using a writing tool with Mansus Grasp active in your offhand. The actual Codex is an unlockable ritual item now.
add: The Living Heart is no longer a roundstart heretic item - their actual heart now becomes their Living Heart, and it makes a sound when triggered. Losing your heart (being disemboweled) will require you to do a ritual to regain it.
add: The Hereic Antag UI has been overhauled, and now hosts much of their mechanics as well as providing some helpful tips for newer players.
add: Most heretic spells now require a focus to cast. All heretics can make a basic focus necklace, and some heretic equipment also functions as a focus. (Credit to Imaginos for the focus sprite!)
add: Heretics now passively gain +1 influence every 20 minutes.
add: Heretic sacrificing has been reworked. You can now sacrifice people who are in soft crit or weaker. Sacrificing someone heals them, cuffs them, and teleports them to the SHADOW REALM, where they must dodge a barrage of hands to survive. Survive long enough and you return without memory - die, and your body will be thrown back.
add: Heretics now have a few new rituals, including the Ritual of Knowledge, a randomly generated ritual that awards knowledge points.
add: Heretic ascension now has a few requirements - you must complete your objectives assigned to you prior to learning the final ritual, and all the final rituals have been changed a bit!
tweak: Used heretic influences can now be removed with a Anomaly Neutralizers.
balance: Some heretic rituals are now limited in the amount they can make. You can only have up to 2 heretic blades crafted at once (3 if you are Path of Flesh).
balance: The Lord of the Night has been buffed to be a little scarier. Did you know the Lord of the Night can eat arms to regain body parts and heal?
balance: Buffed some heretic summons - mostly their health pools.
balance: Nerfed the heretic's Mask of Madness. It can no longer infinite stam-crit you.
balance: Nerfed the heretic's ash mark.
balance: Nerfed a bunch of on-hit-heretic-blade effects. Many effects only apply on mark detonation now: Void blade silence, flesh blade wounds, ash blade gasp cooldown refund.
fix: Fixed quite a few bugs and unintended behaviors with heretic code.
refactor: Refactored and improved much of Heretic code. Improved the file structure dramatically.
admin: The heretic's traitor panel has been beefed up a bit.
refactor: Refactored cult rune inscribing into a component
refactor: Refactored null rod rune clearing into a component
code: Rust is now a component
refactor: refactor almost all of the rust heretic atom procs
add: Wirebrush for advanced rust removal
add: A new heretic node to replace rust wave: entropic plume.
add: A small rune visual effects appears 70% of the time when a tile is rusted.
del: Removed the rust wave node.
fix: The ash man and rust walker nodes now link to the correct nodes.
/:cl:

### CODE REVIEW

File: code/modules/antagonists/heretic/influences.dm [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	if(IS_HERETIC(user))
+		to_chat(user, "<span class='boldwarning'>You know better than to tempt forces out of your control!</span>")
+		return
+
+	var/mob/living/carbon/human/human_user = user
+
+	// A very elaborate way to suicide
+	to_chat(human_user, "<span class='userdanger'>Eldritch energy lashes out, piercing your fragile mind, tearing it to pieces!</span>")
+	human_user.ghostize()
+	var/obj/item/bodypart/head/head = locate() in human_user.bodyparts
```
Comment (Kapu1178) [inline code/modules/antagonists/heretic/influences.dm]:
```suggestion
	var/obj/item/bodypart/head/head = human_user.get_bodypart(BODY_ZONE_HEAD)
```

Comment (Ivniinvi) [reply code/modules/antagonists/heretic/influences.dm]: already fixed lol

**Comment by ToasterBan** [comment]
>Draining influences can be done with "right click".

how exactly does with work? right click as in inside the popup menu?

**Comment by Ivniinvi** [comment]
> > Draining influences can be done with "right click".
> 
> how exactly does with work? right click as in inside the popup menu?

I forgot to remove that from the description lol, still left click.

**Comment by imsynthex** [comment]
> The Living Heart is now the heretic's heart. Literally. It's the heart in their chest. Their heart takes on the appearance of a living heart, and they can pulse it on demand to track their targets. This makes an audible noise.

this could very easily turn into a revs type situation where whenever someone hears the sound all of security gets called over to swarm the place and search everyone, maybe this could be replaced with for example the heretic's eyes going red for a little while after using it as to make it a concealable thing?

**Comment by TeomanTheGreat** [comment]
Holy shit yes. This pr is amazing. Finally heretic will be more interesting and fun as well as less shit.

**Comment by Ivniinvi** [comment]
> > The Living Heart is now the heretic's heart. Literally. It's the heart in their chest. Their heart takes on the appearance of a living heart, and they can pulse it on demand to track their targets. This makes an audible noise.
> 
> this could very easily turn into a revs type situation where whenever someone hears the sound all of security gets called over to swarm the place and search everyone, maybe this could be replaced with for example the heretic's eyes going red for a little while after using it as to make it a concealable thing?

A lot of the time, there just won’t be anything to find with the search. The codex is an optional side path item and the living heart is an organ, so no hard tell items like the previous iteration of heretics.

**Comment by RDS88-beeman** [comment]
> Man was near heartbeat. One of them is valid. Take them all in and check all their chests.

**Comment by Crossedfall** [comment]
Why wasn't this split across multiple PRs? The unrelated heretic ports could have easily been their own things. As it stands, this is ~14,000 changes and nobody is going to want to review this.

**Comment by Crossedfall** [comment]
Talked myself into closing this. Port the unrelated stuff first and then start bringing in the Heretic stuff. Don't do them in bulk, it makes reviewing an insufferable chore.

**Pull request closed without merging**